File: chart/src/main/java/cn/jingzhuan/lib/chart3/renderer/AbstractRenderer.java
package cn.jingzhuan.lib.chart3.renderer;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;

import cn.jingzhuan.lib.chart.Viewport;
import cn.jingzhuan.lib.chart.data.AbstractDataSet;
import cn.jingzhuan.lib.chart.data.ChartData;
import cn.jingzhuan.lib.chart.data.Value;
import cn.jingzhuan.lib.chart3.base.AbstractChartView;

/**
 * @since 2023-09-04
 * created by lei
 */
public abstract class AbstractRenderer<V extends Value, T extends AbstractDataSet<V>> {

    protected Viewport mViewport;

    protected Rect mContentRect;

    protected Paint mRenderPaint;

    protected Paint mLabelTextPaint;

    public AbstractRenderer(AbstractChartView<V, T> chart) {
        this.mViewport = chart.getCurrentViewport();
        this.mContentRect = chart.getContentRect();

        mRenderPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mRenderPaint.setStyle(Paint.Style.STROKE);

        mLabelTextPaint = new Paint();
        mLabelTextPaint.setAntiAlias(true);
    }


    /**
     * Computes the pixel offset for the given X lib value. This may be outside the view bounds.
     */
    protected float getDrawX(float x) {
        return mContentRect.left + mContentRect.width() * (x - mViewport.left) / mViewport.width();
    }

    /**
     * Computes the pixel offset for the given Y lib value. This may be outside the view bounds.
     */
    protected float getDrawY(float y) {
        return mContentRect.bottom - mContentRect.height() * (y - mViewport.top) / mViewport.height();
    }

    public void setTypeface(Typeface tf) {
        mLabelTextPaint.setTypeface(tf);
    }

    public Paint getRenderPaint() {
        return mRenderPaint;
    }

    public ChartData<T> getChartData(){
        return new ChartData<>();
    }

    public abstract void renderer(Canvas canvas);
}

File: chart/src/main/java/cn/jingzhuan/lib/chart3/renderer/AbstractRenderer.kt
package cn.jingzhuan.lib.chart3.renderer

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import cn.jingzhuan.lib.chart3.Viewport
import cn.jingzhuan.lib.chart3.base.AbstractChartView
import cn.jingzhuan.lib.chart3.data.ChartData
import cn.jingzhuan.lib.chart3.data.dataset.AbstractDataSet
import kotlin.math.max

/**
 * @since 2023-09-05
 * @author lei
 */
abstract class AbstractRenderer<T : AbstractDataSet<*>>(chart: AbstractChartView<T>) {
    protected val chartView: AbstractChartView<T>

    protected var currentViewport: Viewport

    protected var contentRect: Rect

    var renderPaint: Paint

    protected var labelTextPaint: Paint

    init {
        this.chartView = chart

        currentViewport = chart.currentViewport
        contentRect = chart.contentRect

        renderPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        renderPaint.style = Paint.Style.STROKE

        labelTextPaint = Paint()
        labelTextPaint.isAntiAlias = true

        this.getChartData()?.setChart(chart)
    }

    open fun getEntryIndex(x: Float): Int {
        val dataSet = getChartData()?.getTouchDataSet() ?: return 0

        val valueCount = max(chartView.totalEntryCount, dataSet.forceValueCount)

        var index: Int = (((x - contentRect.left) * currentViewport.width() / contentRect.width() + currentViewport.left) * valueCount.toFloat()).toInt()
        if (index >= dataSet.values.size) index = dataSet.values.size - 1
        if (index < 0) index = 0

        return index
    }

    open fun getEntryX(index: Int): Float {
        val data = getChartData() ?: return -1f

        val dataSet = data.getTouchDataSet() ?: return -1f

        val valueCount = data.getTouchEntryCount()

        val scale = 1.0f / currentViewport.width()

        val visibleRange = dataSet.getVisibleRange(currentViewport)
        val pointWidth = contentRect.width() / max(visibleRange, dataSet.minValueCount.toFloat())

        val step = contentRect.width() * scale / valueCount
        val startX = contentRect.left - currentViewport.left * contentRect.width() * scale

        var x = startX + step * index + pointWidth * 0.5f

        if (x > contentRect.right) x = contentRect.right.toFloat()
        if (x < contentRect.left) x = contentRect.left.toFloat()

        return x
    }

    /**
     * Computes the pixel offset for the given X lib value. This may be outside the view bounds.
     */
    protected fun getDrawX(x: Float): Float {
        return contentRect.left + contentRect.width() * (x - currentViewport.left) / currentViewport.width()
    }

    /**
     * Computes the pixel offset for the given Y lib value. This may be outside the view bounds.
     */
    protected fun getDrawY(y: Float): Float {
        return contentRect.bottom - contentRect.height() * (y - currentViewport.top) / currentViewport.height()
    }

    open fun setTypeface(tf: Typeface?) {
        labelTextPaint.typeface = tf
    }

    open fun addDataSet(dataSet: T?) {
        getChartData()?.add(dataSet)
    }

    /**
     * 清掉chartData
     */
    open fun clearDataSet(){
        getChartData()?.clear()
    }

    /**
     * 重新计算viewport
     */
    open fun calcDataSetMinMax() {
        getChartData()?.calcMaxMin(currentViewport, contentRect, chartView.offsetPercent)
    }

    open fun getChartData(): ChartData<T>? {
        return null
    }

    abstract fun renderer(canvas: Canvas)
}
